node: Avoid duplicate key constraints when changing chain shard

erayack opened the pull request:
## Description

This PR fixes issue #6196 by implementing a robust backup naming system that prevents duplicate key constraint violations when reverting chain shard changes.

## Problem

When changing a chain's shard using `graphman chain change-shard`, the existing chain is renamed to `<chain>-old`. However, if a subsequent attempt is made to revert the chain's shard back to the original shard, the operation fails due to a unique constraint violation because `<chain>-old` already exists in the database.

## Solution

- **Robust backup naming**: Implemented `next_backup_name()` function that generates unique backup names by appending numeric suffixes when conflicts exist
- **Backup reuse optimization**: When reverting to the original shard, the system now reuses the existing backup instead of creating a new one
- **Backup preservation**: Previous backups are preserved with unique names to maintain data integrity
- **Logging**: Added logging to track backup operations and outcomes

## Changes

- Added `ChainSwapOutcome` struct to track the results of chain swap operations
- Implemented `next_backup_name()` function for conflict-free backup naming
- Enhanced `change_block_cache_shard()` function with robust backup handling logic
- Added detailed logging for backup operations

## Testing

The fix handles the following scenarios:
1. Initial shard change (creates `<chain>-old`)
2. Reverting to original shard (reuses existing backup)
3. Multiple shard changes (preserves all backups with unique names)

Fixes #6196

erayack committed:
node: Fix duplicate key constraint when reverting chain shard changes

- Implement robust backup naming system to avoid conflicts
- Add support for reusing existing backups when appropriate
- Preserve previous backups with unique names
- Add comprehensive logging for backup operations

Fixes #6196

lutter commented:
We really need to review this, this shouldn't have been closed

erayack commented:
Hi @lutter, I rebased PR #6199 onto current master and resolved the conflict in node/src/manager/commands/chain.rs.

The PR is now mergeable. I kept the original fix intent, but ported it to the current async chain-store code, preserved the safe chain-store creation ordering, simplified the backup-name selection flow, and added focused tests for backup-name formatting.

dimitrovmaksim commented on node/src/manager/commands/chain.rs:
```diff
@@ -250,51 +288,112 @@ pub async fn change_block_cache_shard(
 
-            // Update the current chain name to chain-old
-            update_chain_name(conn, &chain_name, &new_name).await?;
+            let previous_backup_final_name = if let Some(backup) = existing_backup.as_ref() {
```
Your code will hit the same constraint in the following case:
mainnet on shard_a
mainnet-old on shard_b

Then running:
`graphman chains change-shard mainnet shard_b`
 
- Line 326 computes the next free backup name, e.g. `mainnet-old-1`.
- Line 327 renames `mainnet-old` -> `mainnet-old-1`.
- Line 330 then tries to rename `mainnet-old-1` -> `mainnet`.

At that point the original `mainnet` row still exists, so the rename to mainnet violates the unique constraint on chains.name. The current `mainnet` -> `mainnet-old-{next}` rename only happens later on lines 339-340.

More generally, I’m not sure preserving multiple old caches is the right approach here. Moving a chain cache between shards should be a rare maintenance operation, and old caches are only useful as an immediate rollback. Reusing an existing backup on the target shard could make sense for an immediate rollback, although in other cases it may be too outdated. For a sequence like A -> B -> C, keeping the old cache on A does not seem useful, so probably dropping that previous backup before preserving the current B cache as the new mainnet-old would make the behavior simpler and avoid having multiple stale caches around the shards.

erayack replied:
I updated the flow to keep only one rollback backup and to order the reuse case as mainnet-old -> temp, mainnet -> mainnet-old, temp -> mainnet so the unique constraint cannot fire. 

One thing I’m less certain about is policy: when an older mainnet-old exists and we are not reusing it, do you want that old backup dropped only after the new change-shard succeeds, or should it be deleted earlier as part of simplifying the behavior? Also, are you okay with updating the in-memory BlockStore mapping explicitly during these renames, or would you prefer a simpler approach even if it leaves cleanup to a reload?

dimitrovmaksim replied:
Sorry for the late reply, I was contemplating what we really want to happen here, so we don't make you go back and forth. What i'm thinking is:
- Add a check if `{chain}-old` already exists and:
  - If its shard matches the target shard, prompt the user that it will re-use that backup (they can decline and remove it with `chain remove` if they want)
  - If target shard is a different shard, print hint to use `chain remove`.

The point is to give more control to the indexers rather than doing stuff for then in the background

erayack commented:
Implemented the behavior you suggested. @dimitrovmaksim 

Current flow is now:

- if {chain}-old exists on a different shard, the command aborts and tells the user to remove it explicitly with graphman chain remove
- if {chain}-old exists on the target shard, the command prompts before reusing it
- otherwise it keeps the current cache as {chain}-old and creates a fresh cache on the target shard

I also changed the reuse ordering to avoid the unique constraint issue: {chain}-old -> temp, {chain} -> {chain}-old, temp -> {chain}

I added focused unit tests for the boundary decision logic.

dimitrovmaksim commented on node/src/manager/commands/chain.rs:
```diff
@@ -250,51 +305,134 @@ pub async fn change_block_cache_shard(
             .execute(conn).await?;
-            Ok(())
+
+            Ok(ChainSwapOutcome {
```
I think this return value is redundant. The transaction does not compute any new state; ChainSwapOutcome just returns `reuse_existing_backup` and `allocated_chain.is_some()`, which are already known before entering the transaction. This can return () and the struct can be completelly removed.

dimitrovmaksim commented on node/src/manager/commands/chain.rs:
```diff
@@ -250,51 +305,134 @@ pub async fn change_block_cache_shard(
         .await
         .ok_or_else(|| anyhow!("unknown chain: {}", &chain_name))?;
-    let new_name = format!("{}-old", &chain_name);
     let ident = chain_store.chain_identifier().await?;
```
One thing I missed is that the current chain and the backup Identities may differ. `graphman` provides a command to update the genesis hash `graphman chain update-genesis`, which is part of the ChainIdentifier, so we should fetch the backup identifier as well and compare them and abort If they are different.

dimitrovmaksim commented on node/src/manager/commands/chain.rs:
```diff
@@ -250,51 +305,134 @@ pub async fn change_block_cache_shard(
+                return Ok(());
+            }
+        }
+    }
```
The `ExistingBackupDisposition` enum seems a little over-modeled for this control flow. It only wraps the None / same shard / wrong shard cases. 
It can be inlide to something like
```rust
let reuse_existing_backup = match existing_backup.as_ref() {
      None => false,
      Some(backup) if backup.shard != target_shard => {
          bail!(...);
      }
      Some(backup) => {
         ...
      }
  };
```
This will remove the need to manage the lifetime on the enum.

dimitrovmaksim commented on node/src/manager/commands/chain.rs:
```diff
@@ -250,51 +305,134 @@ pub async fn change_block_cache_shard(
-
-            // Create a new chain with the name in the destination shard
-            let _ = add_chain(conn, &chain_name, &shard, ident).await?;
+            let temp_backup_name = if let Some(backup) = existing_backup.as_ref() {
```
The temp backup name helpers may be too much for what we need. It's only a transient name for the three-way rename, so something hardcoded like `{chain}-old-temp` should probably be enough.

dimitrovmaksim commented on node/src/manager/commands/chain.rs:
```diff
@@ -250,51 +305,134 @@ pub async fn change_block_cache_shard(
+            update_chain_name(conn, &chain_name, &canonical_backup_name).await?;
+
+            if reuse_existing_backup {
+                debug_assert!(temp_backup_name.is_some());
```
I would not use the `debug_assert!` here, for --release build it will be compiled out, and it will panic on the unwrap() after that anyways.

erayack commented:
Hi @dimitrovmaksim again. Force-pushed 334940b → 704e10d addressing the latest round:

- Removed redundant transaction outcome plumbing (ChainSwapOutcome) and made the transaction return ().
- Inlined the backup disposition control flow and removed the extra enum/helper layering.
 - Added chain-identifier safety checks for backup reuse ({chain}-old must match the active chain identifier), switched transient rename to {chain}-old-temp, and removed the debug_assert!/unwrap() path.

  Tests still pass locally. Ready for another look when you have a moment.

dimitrovmaksim commented on node/src/manager/commands/chain.rs:
```diff
@@ -250,51 +251,138 @@ pub async fn change_block_cache_shard(
+        );
+    }
+
+    if allocated_chain.is_some() {
```
nit: Same as the `allocated_chain = ...` comment. I think this print is redundant.

dimitrovmaksim commented on node/src/manager/commands/chain.rs:
```diff
@@ -250,51 +251,138 @@ pub async fn change_block_cache_shard(
         "Changed block cache shard for {} from {} to {}",
         chain_name, old_shard, shard
     );
+    println!("Latest backup recorded as `{}`", canonical_backup_name);
```
nit: I would remove this println!, seems like a left-over from the sufixed backup names

dimitrovmaksim commented on node/src/manager/commands/chain.rs:
```diff
@@ -250,51 +251,138 @@ pub async fn change_block_cache_shard(
+    };
 
-            let chain = BlockStore::allocate_chain(conn, &chain_name, &shard, &ident).await?;
+    let allocated_chain = if reuse_existing_backup {
```
nit: `allocated_chain` could be boolean, as it is used only for the conditional printing here https://github.com/erayack/graph-node/blob/8b649af8184d6a83e730b4ddb74289c07d3c5cd8/node/src/manager/commands/chain.rs#L380-L385. Tbf I think this message may be redundant as we already notify the user that the chain has been moved to another shard. Personally i would remove that print and change this to 
```
if !reuse_existing_backup {
    let chain = BlockStore::allocate_chain(&mut conn, &chain_name, &target_shard, &ident).await?;
    store.add_chain_store(&chain, true).await?;
}
```

dimitrovmaksim commented on node/src/manager/commands/chain.rs:
```diff
@@ -250,51 +251,138 @@ pub async fn change_block_cache_shard(
         .ok_or_else(|| anyhow!("unknown chain: {}", &chain_name))?;
-    let new_name = format!("{}-old", &chain_name);
     let ident = chain_store.chain_identifier().await?;
+    let target_shard = Shard::new(shard.clone())?;
```
nit: You can drop the `clone()` here and let Shard::new consume the String, then you can use `target_shard` directly later as print arguments instead of `shard`

dimitrovmaksim commented:
Hey @erayack sorry for the delay, I added couple of nit comment. I'll ask another team member to take a look before approving, just to be sure I'm not missing something.

Btw could you drop the merge commit and instead rebase your branch ontop of master? 🙏🏻

erayack commented:
Hi @dimitrovmaksim  Addressed the nits, dropped the merge commit, and rebased onto master. 

Ready for another look whenever you have a moment, and happy for you to loop in the second reviewer in parallel.

erayack committed:
node: Remove duplicate backup store lookup

incrypto32 reviewed (approved):
LGTM

dimitrovmaksim commented:
Hi @erayack, thank you for your contribution! Merging now

erayack commented:
Thanks @dimitrovmaksim and @incrypto32 for the reviews, and thanks @lutter for bringing this back up. Appreciate the guidance on the rollback behavior and cleanup along the way.
